README.md
packages/abitype/README.md

CHANGELOG.md
# abitype

## 0.9.10

### Patch Changes

- [#194](https://github.com/wagmi-dev/abitype/pull/194) [`380c9d9`](https://github.com/wagmi-dev/abitype/commit/380c9d9786f673185369f652f3bde8564b30ff81) Thanks [@mathieu-bour](https://github.com/mathieu-bour)! - Changed the following types to readonly in zod package:

  - `AbiContructor.inputs`
  - `AbiError.inputs`
  - `AbiEvent.inputs`
  - `AbiFunction.inputs` / `AbiFunction.outputs`

## 0.9.9

### Patch Changes

- [#192](https://github.com/wagmi-dev/abitype/pull/192) [`998fc3c`](https://github.com/wagmi-dev/abitype/commit/998fc3c805c75275be509db70eb76b1af8a50411) Thanks [@Raiden1411](https://github.com/Raiden1411)! - Fixed an issue where Zod's parsed ABIs where not assignable to the `Abi` type.

## 0.9.8

### Patch Changes

- [#187](https://github.com/wagmi-dev/abitype/pull/187) [`4a0131f`](https://github.com/wagmi-dev/abitype/commit/4a0131fa505f3d269064ca15f566c2a4adf3a532) Thanks [@tmm](https://github.com/tmm)! - Added `unknown` support to `Narrow`

## 0.9.7

### Patch Changes

- [#185](https://github.com/wagmi-dev/abitype/pull/185) [`f5b27a8`](https://github.com/wagmi-dev/abitype/commit/f5b27a881d9d56aaa632ff55f58a6ed9933965d1) Thanks [@StanleyDing](https://github.com/StanleyDing)! - Fixed a bug where an empty string on the `name` property in AbiParameter wasn't allowed.

## 0.9.6

### Patch Changes

- [#179](https://github.com/wagmi-dev/abitype/pull/179) [`f0f2503`](https://github.com/wagmi-dev/abitype/commit/f0f2503d605cad7cdafe937e35d6b1e00fffef1a) Thanks [@Raiden1411](https://github.com/Raiden1411)! - Fixed a bug on `TypedDataToPrimitiveTypes` where it would create an infinite cyclic type if the `type` was an array type of `keyof TypedData`.

## 0.9.5

### Patch Changes

- [#175](https://github.com/wagmi-dev/abitype/pull/175) [`af5af6b`](https://github.com/wagmi-dev/abitype/commit/af5af6bfb814ee2cec5b22d41ace82242be195f1) Thanks [@Raiden1411](https://github.com/Raiden1411)! - Fixed a bug on `formatAbiParameter` where it would infer incorrectly if the nested elements of the `components` property had no `name` property and the `type` property was of type `tuple` type.

  Before:

  ```ts
  type test = FormatAbiParameter<{
    //   ^? test = "(tuple)"
    type: "tuple";
    components: [{ type: "tuple"; components: [{ type: "string" }] }];
  }>;
  ```

  After:

  ```ts
  type test = FormatAbiParameter<{
    //   ^? type test = "((string))"
    type: "tuple";
    components: [{ type: "tuple"; components: [{ type: "string" }] }];
  }>;
  ```

## 0.9.4

### Patch Changes

- [#176](https://github.com/wagmi-dev/abitype/pull/176) [`bf526d8`](https://github.com/wagmi-dev/abitype/commit/bf526d8775787cba0c472b71ff11c7c7cd143d47) Thanks [@tmm](https://github.com/tmm)! - Fixed issue with format utils where empty name (e.g. `name: ''`) would add extra space after `type`.

## 0.9.3

### Patch Changes

- [#170](https://github.com/wagmi-dev/abitype/pull/170) [`dbe16a2`](https://github.com/wagmi-dev/abitype/commit/dbe16a272c769ede7cd3539795e833b69b785387) Thanks [@Raiden1411](https://github.com/Raiden1411)! - Fixed bug where type instantiation is excessively deep and possibly infinite when using `formatAbiParameter` with `exactOptionalPropertyTypes` set to `false`.

## 0.9.2

### Patch Changes

- [#165](https://github.com/wagmi-dev/abitype/pull/165) [`27b763b`](https://github.com/wagmi-dev/abitype/commit/27b763ba0235675b3d1175de738857c1ba3dbd5f) Thanks [@Raiden1411](https://github.com/Raiden1411)! - Added zod EIP-712 typed data parser.

## 0.9.1

### Patch Changes

- [#164](https://github.com/wagmi-dev/abitype/pull/164) [`ff7a76c`](https://github.com/wagmi-dev/abitype/commit/ff7a76c5a2ce61cb3feee2f5244f4e52d4f6efe1) Thanks [@ardislu](https://github.com/ardislu)! - Fixed human-readable identifiers to allow single leading dollar-sign or underscore, and disallow single leading number.

## 0.9.0

### Minor Changes

- [#159](https://github.com/wagmi-dev/abitype/pull/159) [`267e335`](https://github.com/wagmi-dev/abitype/commit/267e335fc692ca0131a1d60a188f8b340ff7b9ed) Thanks [@frangio](https://github.com/frangio)! - Marked TypeScript as optional peer dependency

### Patch Changes

- [#162](https://github.com/wagmi-dev/abitype/pull/162) [`454cf87`](https://github.com/wagmi-dev/abitype/commit/454cf87643a99944fe070dbc850cbee344b49f01) Thanks [@tmm](https://github.com/tmm)! - Added state mutability filter to `ExtractAbiFunction`.

## 0.8.11

### Patch Changes

- [#156](https://github.com/wagmi-dev/abitype/pull/156) [`d577f3d`](https://github.com/wagmi-dev/abitype/commit/d577f3d533b2ef4477a4023ee077ffaf2b6a76d0) Thanks [@tmm](https://github.com/tmm)! - Switched `TypedData` from TypeScript object to Record.

## 0.8.10

### Patch Changes

- [`1d62393`](https://github.com/wagmi-dev/abitype/commit/1d62393d97ca927feeec8eacb4588952fd635ec0) Thanks [@tmm](https://github.com/tmm)! - Fixed `parseAbi` issue with `typescript@5.1.3`

## 0.8.9

### Patch Changes

- [#151](https://github.com/wagmi-dev/abitype/pull/151) [`482bc5f`](https://github.com/wagmi-dev/abitype/commit/482bc5f2258fb60fe52f252be599c98ef753b24c) Thanks [@tmm](https://github.com/tmm)! - Added human-readble ABI format utils.

## 0.8.8

### Patch Changes

- [#149](https://github.com/wagmi-dev/abitype/pull/149) [`ba21aae`](https://github.com/wagmi-dev/abitype/commit/ba21aae2460efe4ee6dae074c85e73d6c87a81d4) Thanks [@tmm](https://github.com/tmm)! - Removed internal barrel exports.

## 0.8.7

### Patch Changes

- [#147](https://github.com/wagmi-dev/abitype/pull/147) [`f9aa86e`](https://github.com/wagmi-dev/abitype/commit/f9aa86ea549bbefc471ec9921b6e8cb6ebfd985f) Thanks [@Raiden1411](https://github.com/Raiden1411)! - Fixed type-level issue with human-readable function signatures with no space returns.

## 0.8.6

### Patch Changes


src/abi.ts
import type { ResolvedConfig } from './config.js'
import type { Pretty, Range } from './types.js'

export type Address = ResolvedConfig['AddressType']

////////////////////////////////////////////////////////////////////////////////////////////////////
// Solidity Types

// Could use `Range`, but listed out for zero overhead
// rome-ignore format: no formatting
export type MBytes =
  | '' | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
  | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19
  | 20 | 21 | 22 | 23 | 24 | 25 | 26 | 27 | 28 | 29
  | 30 | 31 | 32
// rome-ignore format: no formatting
export type MBits =
  | '' | 8 | 16 | 24 | 32 | 40 | 48 | 56 | 64 | 72
  | 80 | 88 | 96 | 104 | 112 | 120 | 128 | 136 | 144 | 152
  | 160 | 168 | 176 | 184 | 192 | 200 | 208 | 216 | 224 | 232
  | 240 | 248 | 256

// From https://docs.soliditylang.org/en/latest/abi-spec.html#types
export type SolidityAddress = 'address'
export type SolidityBool = 'bool'
export type SolidityBytes = `bytes${MBytes}` // `bytes<M>`: binary type of `M` bytes, `0 < M <= 32`
export type SolidityFunction = 'function'
export type SolidityString = 'string'
export type SolidityTuple = 'tuple'
export type SolidityInt = `${'u' | ''}int${MBits}` // `(u)int<M>`: (un)signed integer type of `M` bits, `0 < M <= 256`, `M % 8 == 0`
// No need to support "fixed" until Solidity does
// https://github.com/ethereum/solidity/issues/409
// `(u)fixed<M>x<N>`: (un)signed fixed-point decimal number of `M` bits, `8 <= M <= 256`, `M % 8 == 0`,
// and `0 < N <= 80`, which denotes the value `v` as `v / (10 ** N)`
// export type SolidityFixed =
//   | `${'u' | ''}fixed`
//   | `${'u' | ''}fixed${MBits}x${Range<1, 20>[number]}`

export type SolidityFixedArrayRange = Range<
  ResolvedConfig['FixedArrayMinLength'],
  ResolvedConfig['FixedArrayMaxLength']
>[number]
export type SolidityFixedArraySizeLookup = {
  [Prop in SolidityFixedArrayRange as `${Prop}`]: Prop
}

/**
 * Recursively build arrays up to maximum depth
 * or use a more broad type when maximum depth is switched "off"
 */
type _BuildArrayTypes<
  T extends string,
  Depth extends readonly number[] = [],
> = ResolvedConfig['ArrayMaxDepth'] extends false
  ? `${T}[${string}]`
  : Depth['length'] extends ResolvedConfig['ArrayMaxDepth']
  ? T
  : T extends `${any}[${SolidityFixedArrayRange | ''}]`
  ? _BuildArrayTypes<T | `${T}[${SolidityFixedArrayRange | ''}]`, [...Depth, 1]>
  : _BuildArrayTypes<`${T}[${SolidityFixedArrayRange | ''}]`, [...Depth, 1]>

// Modeling fixed-length (`<type>[M]`) and dynamic (`<type>[]`) arrays
// Tuple and non-tuple versions are separated out for narrowing anywhere structs show up
export type SolidityArrayWithoutTuple = _BuildArrayTypes<
  | SolidityAddress
  | SolidityBool
  | SolidityBytes
  | SolidityFunction
  | SolidityInt
  | SolidityString
>
export type SolidityArrayWithTuple = _BuildArrayTypes<SolidityTuple>
export type SolidityArray = SolidityArrayWithoutTuple | SolidityArrayWithTuple

////////////////////////////////////////////////////////////////////////////////////////////////////
// Abi Types

export type AbiType =
  | SolidityArray
  | SolidityAddress
  | SolidityBool
  | SolidityBytes
  | SolidityFunction
  | SolidityInt
  | SolidityString
  | SolidityTuple
type ResolvedAbiType = ResolvedConfig['StrictAbiType'] extends true
  ? AbiType
  : string

export type AbiInternalType =
  | ResolvedAbiType
  | `address ${string}`
  | `contract ${string}`
  | `enum ${string}`
  | `struct ${string}`

export type AbiParameter = Pretty<
  {
    type: ResolvedAbiType
    name?: string | undefined
    /** Representation used by Solidity compiler */
    internalType?: AbiInternalType | undefined
  } & (
    | { type: Exclude<ResolvedAbiType, SolidityTuple | SolidityArrayWithTuple> }
    | {
        type: SolidityTuple | SolidityArrayWithTuple
        components: readonly AbiParameter[]
      }
  )
>

export type AbiEventParameter = Pretty<
  AbiParameter & { indexed?: boolean | undefined }
>

/**
 * State mutability for {@link AbiFunction}
 *
 * @see https://docs.soliditylang.org/en/latest/contracts.html#state-mutability
 */
export type AbiStateMutability = 'pure' | 'view' | 'nonpayable' | 'payable'

/** Kind of {@link AbiParameter} */
export type AbiParameterKind = 'inputs' | 'outputs'

/** ABI ["function"](https://docs.soliditylang.org/en/latest/abi-spec.html#json) type */
export type AbiFunction = {
  type: 'function'
  /**
   * @deprecated use `pure` or `view` from {@link AbiStateMutability} instead
   * @see https://github.com/ethereum/solidity/issues/992
   */
  constant?: boolean | undefined
  /**
   * @deprecated Vyper used to provide gas estimates
   * @see https://github.com/vyperlang/vyper/issues/2151
   */
  gas?: number | undefined
  inputs: readonly AbiParameter[]
  name: string
  outputs: readonly AbiParameter[]
  /**
   * @deprecated use `payable` or `nonpayable` from {@link AbiStateMutability} instead
   * @see https://github.com/ethereum/solidity/issues/992
   */
  payable?: boolean | undefined
  stateMutability: AbiStateMutability
}

/** ABI ["constructor"](https://docs.soliditylang.org/en/latest/abi-spec.html#json) type */
export type AbiConstructor = {
  type: 'constructor'
  inputs: readonly AbiParameter[]
  /**
   * @deprecated use `payable` or `nonpayable` from {@link AbiStateMutability} instead
   * @see https://github.com/ethereum/solidity/issues/992
   */
  payable?: boolean | undefined
  stateMutability: Extract<AbiStateMutability, 'payable' | 'nonpayable'>
}

/** ABI ["fallback"](https://docs.soliditylang.org/en/latest/abi-spec.html#json) type */
export type AbiFallback = {
  type: 'fallback'
  inputs?: readonly [] | undefined
  /**
   * @deprecated use `payable` or `nonpayable` from {@link AbiStateMutability} instead
   * @see https://github.com/ethereum/solidity/issues/992
   */
  payable?: boolean | undefined
  stateMutability: Extract<AbiStateMutability, 'payable' | 'nonpayable'>
}

/** ABI ["receive"](https://docs.soliditylang.org/en/latest/contracts.html#receive-ether-function) type */
export type AbiReceive = {
  type: 'receive'
  stateMutability: Extract<AbiStateMutability, 'payable'>
}

/** ABI ["event"](https://docs.soliditylang.org/en/latest/abi-spec.html#events) type */
export type AbiEvent = {
  type: 'event'
  anonymous?: boolean | undefined
  inputs: readonly AbiEventParameter[]
  name: string
}

/** ABI ["error"](https://docs.soliditylang.org/en/latest/abi-spec.html#errors) type */
export type AbiError = {
  type: 'error'
  inputs: readonly AbiParameter[]
  name: string
}

/** `"type"` name for {@link Abi} items. */
export type AbiItemType =
  | 'constructor'
  | 'error'
  | 'event'
  | 'fallback'
  | 'function'
  | 'receive'

/**
 * Contract [ABI Specification](https://docs.soliditylang.org/en/latest/abi-spec.html#json)
 */
export type Abi = readonly (
  | AbiConstructor
  | AbiError
  | AbiEvent
  | AbiFallback
  | AbiFunction
  | AbiReceive
)[]

////////////////////////////////////////////////////////////////////////////////////////////////////
// Typed Data Types

export type TypedDataDomain = {
  chainId?: number | undefined
  name?: string | undefined
  salt?: ResolvedConfig['BytesType']['outputs'] | undefined
  verifyingContract?: Address | undefined
  version?: string | undefined
}

// Subset of `AbiType` that excludes `tuple`, `function`, and dynamic aliases `int` and `uint`
export type TypedDataType = Exclude<
  AbiType,
  SolidityFunction | SolidityTuple | SolidityArrayWithTuple | 'int' | 'uint'
>

export type TypedDataParameter = {
  name: string
  type: TypedDataType | keyof TypedData | `${keyof TypedData}[${string | ''}]`
}

/**
 * [EIP-712](https://eips.ethereum.org/EIPS/eip-712#definition-of-typed-structured-data-%F0%9D%95%8A) Typed Data Specification
 */
export type TypedData = Pretty<
  Record<string, readonly TypedDataParameter[]> & {
    // Disallow `TypedDataType` as key names (e.g. `address`)
    [_ in TypedDataType]?: never
  }
>

export type ResolvedTypedData = {
  [x: string]: TypedDataType | ResolvedTypedData | ResolvedTypedData[]
}
